waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -222,39 +224,49 @@
    - it has restarted too many times (reached the restart-expiry-threshold threshold)
    - the primary container (waiter-apps) has not transitioned to running state in container-running-grace-secs seconds
    - the pod has the waiter/pod-expired=true annotation."
-  [{:keys [container-running-grace-secs restart-expiry-threshold]}
-   instance-id restart-count {:keys [waiter/pod-expired]} primary-container-status pod-started-at]
-  (cond
-    (>= restart-count restart-expiry-threshold)
-    (do
-      (log/info "instance expired as it reached the restart threshold"
-                {:instance-id instance-id
-                 :restart-count restart-count})
-      true)
-    (and pod-started-at
-         (pos? container-running-grace-secs)
-         (empty? (:lastState primary-container-status))
-         (not (contains? (:state primary-container-status) :running))
-         (<= container-running-grace-secs (t/in-seconds (t/interval pod-started-at (t/now)))))
-    (do
-      (log/info "instance expired as it took too long to transition to running state"
-                {:instance-id instance-id
-                 :primary-container-status primary-container-status
-                 :started-at pod-started-at})
-      true)
-    (= "true" pod-expired)
-    (do
-      (log/info "instance expired based on pod annotation"
-                {:instance-id instance-id
-                 :waiter/pod-expired pod-expired})
-      true)))
+  [{:keys [container-running-grace-secs restart-expiry-threshold watch-state]}
+   service-id instance-id restart-count {:keys [waiter/pod-expired waiter/revision-timestamp]}
+   primary-container-status pod-started-at]
+  (let [rs-revision-timestamp (get-in @watch-state [:service-id->service service-id :k8s/revision-timestamp])]
+    (cond
+      (>= restart-count restart-expiry-threshold)
+      (do
+        (log/info "instance expired as it reached the restart threshold"
+                  {:instance-id instance-id
+                   :restart-count restart-count})
+        true)
+      (and pod-started-at
+           (pos? container-running-grace-secs)
+           (empty? (:lastState primary-container-status))
+           (not (contains? (:state primary-container-status) :running))
+           (<= container-running-grace-secs (t/in-seconds (t/interval pod-started-at (t/now)))))
+      (do
+        (log/info "instance expired as it took too long to transition to running state"
+                  {:instance-id instance-id
+                   :primary-container-status primary-container-status
+                   :started-at pod-started-at})
+        true)
+      (= "true" pod-expired)
+      (do
+        (log/info "instance expired based on pod annotation"
+                  {:instance-id instance-id
+                   :waiter/pod-expired pod-expired})
+        true)
+
+      (pos? (compare rs-revision-timestamp revision-timestamp))
```

Comment:
I thought through all of the different combinations of `nil` that are possible in the `compare` here, and I think it works correctly in all cases! 👍

Reply:
I'm also glad to see that you covered all of those cases in the unit tests.

Reply:
Thanks for cross-checking. Spent some time thinking about the cases too and using as simple an expression as possible, happy that compare handles `nil` :)